### PR TITLE
Add TimeoutAttribute

### PR DESCRIPTION
**Pull request by Kingmidas74**

### CODE REVIEW

File: src/NClient/NClient.Abstractions/Annotations/Operations/ITimeoutAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+{
+    public interface ITimeoutAttribute
+    {
+        /// <summary>
+        /// Gets the route order. The order determines the order of route execution. Routes with a lower
+        /// order value are tried first. When a route doesn't specify a value, it gets a default value of 0.
+        /// A null value for the Order property means that the user didn't specify an explicit order for the
+        /// route.
+        /// </summary>
+        int Order { get; set; }
```
Comment (smolchanovsky) [inline src/NClient/NClient.Abstractions/Annotations/Operations/ITimeoutAttribute.cs:13]: Why does the timeout need an order?

File: src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+        private TimeoutAttribute? Find(MethodInfo method)
+        {
+            var attributes = method
+                .GetCustomAttributes()
+                .Select(x => _attributeMapper.TryMap(x))
+                .ToArray();
+            if (attributes.Any(x => x is IPathAttribute))
+                throw _clientValidationExceptionFactory.MethodAttributeNotSupported(nameof(IPathAttribute));
+///WTF
```
Comment (smolchanovsky) [inline src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs:41]: 😄 What's wrong?

File: src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        private TimeoutAttribute? Find(MethodInfo method)
+        {
+            var attributes = method
+                .GetCustomAttributes()
+                .Select(x => _attributeMapper.TryMap(x))
+                .ToArray();
+            if (attributes.Any(x => x is IPathAttribute))
+                throw _clientValidationExceptionFactory.MethodAttributeNotSupported(nameof(IPathAttribute));
```
Comment (smolchanovsky) [inline src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs:40]: This is not required here, the paths are checked in another provider

File: src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+using System.Reflection;
+using NClient.Annotations;
+using NClient.Core.Mappers;
+using NClient.Standalone.Exceptions.Factories;
+
+namespace NClient.Standalone.ClientProxy.Generation.MethodBuilders.Providers
+{
+    internal interface ITimeoutAttributeProvider
+    {
+        ITimeoutAttribute? Get(MethodInfo method, IEnumerable<MethodInfo> overridingMethods);
```
Comment (smolchanovsky) [inline src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs:12]: There is an implicit convention - if a method has a possibility of not finding an attribute and returning null, then it should have the `Find` name.

**Review by smolchanovsky — CHANGES_REQUESTED** [review]
The timeout is now passed in the `Method` class, which is nice. Now you need to pass this timeout to `IRequest`, and then process it in transport. HttpClient does not know how to set timeouts for a request, so you need to use a cancellation token.

File: src/NClient/NClient.Abstractions/Annotations/Operations/ITimeoutAttribute.cs (hunk 1 above)
Comment (Kingmidas74) [reply src/NClient/NClient.Abstractions/Annotations/Operations/ITimeoutAttribute.cs]: Agree. Removed.

File: src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs (hunk 2 above)
Comment (Kingmidas74) [reply src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs]: Legacy.

File: src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs (hunk 3 above)
Comment (Kingmidas74) [reply src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs]: Yep. Removed

File: src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs (hunk 4 above)
Comment (Kingmidas74) [reply src/NClient/NClient.Standalone/ClientProxy/Generation/MethodBuilders/Providers/TimeoutAttributeProvider.cs]: Ok.

File: tests/NClient/NClient.Tests/ClientTests/TimeoutClientTest.cs [hunk 5]
```diff
@@ -46,11 +47,11 @@ public async Task TimeoutClient_GetAsyncWithHttpClientTimeout_ThrowTaskCanceledE
 
             await nclient.Invoking(x => x.GetAsync(id))
                 .Should()
-                .ThrowExactlyAsync<TaskCanceledException>();
+                .ThrowExactlyAsync<OperationCanceledException>();
```
Comment (smolchanovsky) [inline tests/NClient/NClient.Tests/ClientTests/TimeoutClientTest.cs:50]: Why did you change the exception type? Now the tests don't pass: 
> Expected type to be System.OperationCanceledException, but found System.Threading.Tasks.TaskCanceledException.

File: tests/NClient/NClient.Standalone.Tests/Clients/ITimeoutClient.cs [hunk 6]
```diff
@@ -6,12 +6,15 @@
...
     [Path("api/timeout")]
+    [Timeout(2)]
     public interface ITimeoutClientWithMetadata : ITimeoutClient
     {
         [GetMethod]
+        [Timeout(1)]
         new int Get(int id);
         
         [GetMethod]
+        [Timeout(1)]
```
Comment (smolchanovsky) [inline tests/NClient/NClient.Standalone.Tests/Clients/ITimeoutClient.cs:17]: To separate the tested functionality, it is necessary to have two pairs of methods: with an attribute and without it.

File: tests/NClient/NClient.Tests/ClientTests/TimeoutClientTest.cs [hunk 7]
```diff
@@ -14,11 +15,11 @@ namespace NClient.Tests.ClientTests
     public class TimeoutClientTest
     {
         [Test]
-        public void TimeoutClient_GetWithHttpClientTimeout_ThrowTaskCanceledException()
+        public void TimeoutClient_GetWithHttpClientTimeout_ThrowOperationCanceledException()
         {
             const int id = 1;
             using var api = TimeoutApiMockFactory.MockGetMethod(id);
-            var httpClient = new HttpClient { Timeout = 1.Microseconds() };
```
Comment (smolchanovsky) [inline tests/NClient/NClient.Tests/ClientTests/TimeoutClientTest.cs:21]: This is one of the valid methods to set a timeout. I think we should leave this test in its original form.

File: src/NClient.Providers/NClient.Providers.Transport.SystemNetHttp/SystemNetHttpTransportRequestBuilder.cs [hunk 8]
```diff
@@ -32,9 +32,6 @@ public Task<HttpRequestMessage> BuildAsync(IRequest request, CancellationToken c
                 RequestUri = uri
             };
 
-            if (request.Timeout.HasValue)
-                httpRequestMessage.SetTimeout(request.Timeout.Value);
```
Comment (smolchanovsky) [inline src/NClient.Providers/NClient.Providers.Transport.SystemNetHttp/SystemNetHttpTransportRequestBuilder.cs:36]: You should delete this extension, because now it is not used anywhere.

**Review by smolchanovsky — CHANGES_REQUESTED** [review]

File: tests/NClient/NClient.Tests/ClientTests/TimeoutClientTest.cs (hunk 5 above)
Comment (Kingmidas74) [reply tests/NClient/NClient.Tests/ClientTests/TimeoutClientTest.cs:51]: There are some complex questions...

File: tests/NClient/NClient.Standalone.Tests/Clients/ITimeoutClient.cs (hunk 6 above)
Comment (Kingmidas74) [reply tests/NClient/NClient.Standalone.Tests/Clients/ITimeoutClient.cs]: Fix

File: tests/NClient/NClient.Tests/ClientTests/TimeoutClientTest.cs (hunk 7 above)
Comment (Kingmidas74) [reply tests/NClient/NClient.Tests/ClientTests/TimeoutClientTest.cs]: Fix

File: src/NClient.Providers/NClient.Providers.Transport.SystemNetHttp/SystemNetHttpTransportRequestBuilder.cs (hunk 8 above)
Comment (Kingmidas74) [reply src/NClient.Providers/NClient.Providers.Transport.SystemNetHttp/SystemNetHttpTransportRequestBuilder.cs:36]: Dropped

**Comment by Kingmidas74** [comment]
Test

**Review by smolchanovsky — APPROVED** [review]

**Review by Kingmidas74 — COMMENTED** [review]
Good

**Review by smolchanovsky — APPROVED** [review]
